protocols/kad/src/behaviour.rs

```diff
@@ -2036,6 +2074,15 @@ where
         }
     }
 
+    fn handle_bootstrap(&mut self) {
+        if self.initial_bootstrap.is_none() {
+            match self.bootstrap() {
+                Ok(query_id) => self.initial_bootstrap = Some(InitialBootstrap::Active(query_id)),
+                Err(_) => self.initial_bootstrap = None,
+            }
+        }
+    }
```

Comment:
This isn't quite the functionality we want I think!

I think the state we want to store is: "Are we currently running an automated bootstrap?". If we don't then we can kick one of via two means:

1. Because we connected to a new peer
2. Because the timer fired

I think the state I'd store is `Option<QueryId>`. In addition, we can store a state:

```rust
any_bootstrap_successful: bool
```

If this field is `true`, we don't kick of bootstraps for new peers in the routing table. That field we can check before we call this function and then we can name it:

```suggestion
    fn maybe_bootstrap(&mut self) {
        if self.current_automated_bootstrap.is_none() {
            match self.bootstrap() {
                Ok(query_id) => self.initial_bootstrap = Some(query_id),
                Err(_) => self.initial_bootstrap = None,
            }
        }
    }
```

Reply:
In `maybe_bootstrap` in the `Ok(query_id)` case shouldn't I set also that `self.any_bootstrap_successful = true;` ?

Reply:
Calling `bootstrap` is just the start of the query! Whether the query is actually successful can only be observed by looking at the `Event`s that are emitted.

I think for now, we can define "bootstrap successful" as having found at least one other peer via the bootstrap. I'd like @mxinden to confirm this but I think we achieve this by doing:

- Looking at every `Event::OutboundQueryProgressed` event
- Checking if it is `QueryResult::Bootstrap`
- Checking if it is `Ok`

Once we encounter one of these, we can set the `any_bootstrap_successful` to true.

Reply:
Isn't this happening in the `query_finished` function?

Reply:
> Isn't this happening in the `query_finished` function?

You could handle the above in `query_finished`, yes.

Reply:
Now I am setting `any_bootstrap_successful` in `query_finished `